internal/compiler/js_diagnostics.go

```diff
@@ -0,0 +1,527 @@
+package compiler
+
+import (
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/binder"
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/diagnostics"
+	"github.com/microsoft/typescript-go/internal/scanner"
+)
+
+// jsDiagnosticsVisitor is used to find TypeScript-only constructs in JavaScript files
+type jsDiagnosticsVisitor struct {
+	sourceFile  *ast.SourceFile
+	diagnostics []*ast.Diagnostic
+}
+
+// getJSSyntacticDiagnosticsForFile returns diagnostics for TypeScript-only constructs in JavaScript files
+func getJSSyntacticDiagnosticsForFile(sourceFile *ast.SourceFile) []*ast.Diagnostic {
+	visitor := &jsDiagnosticsVisitor{
+		sourceFile:  sourceFile,
+		diagnostics: []*ast.Diagnostic{},
+	}
+
+	// Walk the entire AST to find TypeScript-only constructs
+	visitor.walkNodeForJSDiagnostics(sourceFile.AsNode(), sourceFile.AsNode())
+
+	return visitor.diagnostics
+}
+
+// getJSSyntacticDiagnosticsForFile returns diagnostics for TypeScript-only constructs in JavaScript files
+func (p *Program) getJSSyntacticDiagnosticsForFile(sourceFile *ast.SourceFile) []*ast.Diagnostic {
+	diagnostics, _ := p.jsDiagnosticCache.LoadOrStore(sourceFile, getJSSyntacticDiagnosticsForFile(sourceFile))
+	return diagnostics
```

Comment:
You were too aggressive; you should have kept the `Load`, since this now always walks the ast. You need:


```suggestion
	if cached, ok := p.jsDiagnosticCache.Load(sourceFile); ok {
		return cached
	}
	diagnostics, _ := p.jsDiagnosticCache.LoadOrStore(sourceFile, getJSSyntacticDiagnosticsForFile(sourceFile))
	return diagnostics
```